Reconnect same host plugin process and client

amiramw opened the pull request:
Fixes problem of plugins not working after short disconnection.
Keep disconnected host plugin process alive for at least one minute and
reconnect it if the client is the same.

Signed-off-by: Amiram Wingarten <amiram.wingarten@sap.com>

akosyakov commented on packages/plugin-ext/src/hosted/browser/hosted-plugin-watcher.ts:
```diff
@@ -30,7 +30,9 @@ export class HostedPluginWatcher {
         const logEmitter = this.onLogMessage;
+        const clientId = Math.floor(Math.random() * 1000000) + 1;
         return {
+            getClientId: () => Promise.resolve(clientId),
```
Should not server create client handles then? It knows everything about them.

amiramw replied:
@akosyakov could you point me to some example where client handles are used?

akosyakov replied:
I mean server can give a client an id, since it knows all clients and can deduplicate them. Instead of relying on random numbers.

amiramw replied:
You mean I should add to `HostedPluginClient` a method `setClientId` that the server populates instead of populating it in the client?

amiramw replied:
ok. if i understood correctly then the code is changed now.

benoitf commented on packages/plugin-ext/src/hosted/node/hosted-plugin-process.ts:
```diff
@@ -46,15 +48,39 @@ export class HostedPluginProcess implements ServerPluginRunner {
     private childProcess: cp.ChildProcess | undefined;
+
+    // child processes are kept for one minute in order to reuse them in case of network disconnections
+    private static cachedCPMap: { [clientId: string]:  {cp: cp.ChildProcess, toBeKilledAfter: number}; } = {};
```
hello, question, why static for the map and the clienID ? I'm not sure it's shared at some point

amiramw replied:
The scope of this class instance is one per connection between server and browser.
The whole module in packages/plugin-ext/src/hosted/node/plugin-ext-hosted-backend-module.ts is `ConnectionContainerModule`.
Each reconnect is with new instance and also different browser tabs will reconnect to different instances of this class as far as I experience.

akosyakov replied:
static properties are bad because they live outside of DI context, it would be better to have a singleton instance created from the application DI container that it is shared between all connections and inject it here.

akosyakov replied:
but wait if this class is a per connection then why do you need an id, a map and a seq at all? There can be only one client per a connection, so you should take care only about this client. Each instance of `HostedPluginProcess` will take care about own client.

amiramw replied:
As far as I can see upon connection (new or reconnect) the same client is assigned new server and the old one is disposed.

https://github.com/theia-ide/theia/blob/843c7a0923e07d0c2b7a2f0abe04618c4b69ab08/packages/plugin-ext/src/hosted/node/plugin-ext-hosted-backend-module.ts#L49-L52

Even more than that, the DI container in which this code is run is `ConnectionContainerModule`. It is recreated on connection.

That's why I was forced to use static. If there is a way to use singleton fron another DI container that is not recreated on each reconnect I would prefer that. Or is there a simpler solution for that?

akosyakov replied:
understood, let's just extract a new class which keep mapping and inject it here instead of using static properties

akosyakov commented on packages/plugin-ext/src/hosted/node/hosted-plugin-processes-cache.ts:
```diff
@@ -0,0 +1,64 @@
+/********************************************************************************
+ * Copyright (C) 2019 TypeFox and others.
```
it should be you or entity which you represent, not `TypeFox`

akosyakov commented on packages/plugin-ext/src/hosted/node/hosted-plugin-processes-cache.ts:
```diff
@@ -0,0 +1,64 @@
+export class HostedPluginProcessesCache {
+
+    // child processes are kept for one minute in order to reuse them in case of network disconnections
+    private cachedCPMap: { [clientId: string]: { cp: cp.ChildProcess, toBeKilledAfter: number }; } = {};
```
please use `Map<string, { ... }>` instead

akosyakov commented:
We could accept it as a temporary fix. I don't like it much since it will cause other issue, i.e. you close a window but process tree is hanging for one minutes occupying different ports and so on. So if someone open a new window within one minute and tries to launch something it is not going to work. Maybe timeout can be smaller like 5-10s. @amiramw for how long you have connection issues?

amiramw commented:
@akosyakov We are facing disconnections that take few seconds. I reduced timeout to 10 seconds and added an ENV variable so that we can override it if needed. If you want I can make it even lower. Also I fixed the other code review comments.
Even as temporary solution for 0.7.0 this would be a major improvement in theia productivity for us.
I think that for few seconds of disconnection it is good not to be forced to execute all plugins and vscode extensions again.
For later version I can take a look into a full solution that will work also after the timeout and also after theia restart.

akosyakov commented on packages/plugin-ext/src/hosted/node/hosted-plugin-process.ts:
```diff
@@ -41,20 +42,41 @@ export class HostedPluginProcess implements ServerPluginRunner {
     private client: HostedPluginClient;
 
+    private async getClientId(): Promise<number> {
+        let clientId = await this.client.getClientId();
```
if `clientId` is greater than `clientIdSeq` we have to reject it to support the case when a server is restarted, otherwise they can be several clients with the same id

akosyakov commented:
> I think that for few seconds of disconnection it is good not to be forced to execute all plugins and vscode extensions again.

It is error prone. If a ws connection is gone then some data can be lost which should have been sent, like text edits, so state of plugin on backend is different from frontend state leading to all kind of functionality bugs. If you want to really reuse backend plugins you should make sure that they sync state with frontend when a client reconnects. It is not an easy task as well.

Usually we preserve ws connection for 30s on network issues that data are delivered eventually. In this case it's possible to reuse plugins. If a ws connection is not preserved the easiest strategy is to kill a plugin and start a new one to make sure that it is in sync with frontend state.

On your network, apparently, somehow ws connection is got killed on frontend or backend too early by network errors.

amiramw commented:
@akosyakov I see your point. Maybe timeout small enough can make these inconsistencies pretty rare. 
We have an issue with the way nginx ingress controller handles route changes which cause short disconnection.
It can be reproduced locally with proxy server copied from here: https://github.com/theia-ide/theia/issues/4590. Terminate host plugin process is triggered immediately when killing the proxy.

Anyway as a temporary solution for 0.7.0 I would appreciate if this could be merged. I'll try to investigate in parallel the second approach.

amiramw commented:
@akosyakov @benoitf  If this is acceptable as temporary solution I would be happy if you could merge it before 0.7.0 release today :pray:
